commit 846b5756d7632523b5bfce78c163aa883aa9d587
Author: Chuck Lever <chuck.lever@oracle.com>
Date:   Mon Jan 2 12:05:43 2023 -0500

    SUNRPC: Add an XDR decoding helper for struct opaque_auth
    
    RFC 5531 defines the body of an RPC Call message like this:
    
            struct call_body {
                    unsigned int rpcvers;
                    unsigned int prog;
                    unsigned int vers;
                    unsigned int proc;
                    opaque_auth cred;
                    opaque_auth verf;
                    /* procedure-specific parameters start here */
            };
    
    In the current server code, decoding a struct opaque_auth type is
    open-coded in several places, and is thus difficult to harden
    everywhere.
    
    Introduce a helper for decoding an opaque_auth within the context
    of a xdr_stream. This helper can be shared with all authentication
    flavor implemenations, even on the client-side.
    
    Done as part of hardening the server-side RPC header decoding paths.
    
    Reviewed-by: Jeff Layton <jlayton@kernel.org>
    Signed-off-by: Chuck Lever <chuck.lever@oracle.com>

diff --git a/include/linux/sunrpc/xdr.h b/include/linux/sunrpc/xdr.h
index f84e2a1358e1..8b5c9d0cdcb5 100644
--- a/include/linux/sunrpc/xdr.h
+++ b/include/linux/sunrpc/xdr.h
@@ -346,6 +346,9 @@ ssize_t xdr_stream_decode_string(struct xdr_stream *xdr, char *str,
 		size_t size);
 ssize_t xdr_stream_decode_string_dup(struct xdr_stream *xdr, char **str,
 		size_t maxlen, gfp_t gfp_flags);
+ssize_t xdr_stream_decode_opaque_auth(struct xdr_stream *xdr, u32 *flavor,
+		void **body, unsigned int *body_len);
+
 /**
  * xdr_align_size - Calculate padded size of an object
  * @n: Size of an object being XDR encoded (in bytes)